### ISSUE TITLE
Approach for automating Agent install on mac is unclear

### ISSUE THREAD

**Issue by ColinDabritzViewpoint** [Microsoft/vsts-agent#697]
## **Agent version and platform**
Version of your agent? 
-Attempting to create script to install latest.
<br>
OS of the machine running the agent?
-OSX
<br>

## **VSTS type and version**
VisualStudio.com or On-Prem TFS?
-VSTS on VisualStudio.com
<br>
If VisualStudio.com, what is your account name? 
- http://viewpointvso.visualstudio.com
<br>

## **What's not working?** 

I am attempting to automate the setup of the OSX Vsts-Agent, so new agents can be easily provisioned into a pool of Mac agents. Much of the script is straight forward transcription of the install instructions. The actual agent download of the latest version is giving me trouble.

This may be a matter of documentation, did I miss a documentation resource for this process?

The current, acute problem I am having is deciding how to download the agent release. From the VSTS download page I copied the current URL: `https://github.com/Microsoft/vsts-agent/releases/download/v2.109.1/vsts-agent-osx.10.11-x64-2.109.1.tar.gz`

I can download the URL with curl as shown in the script below.

Is there some URL that would better represent 'latest stable version'? Especially one that would ensure current VSTS compatability?

Is it ok to continue installing an older version for some time, considering the automatic update process should give it the latest agent relatively quickly?

My next challenges are generating the configuration, which looks like it's all in the .agent file, and figuring out how to handle authentication secrets for the first time. The agent appears to remember after that.

I'm not certain that with the existing tools I can fully provision a mac, for one thing I need XCode and Xamarin studio, and XCode appears challenging to install from the command line, but I would like to automate as much of the process as possible to make spinning up additional agents as easy as possible.

Is there currently any agent automated install documentation? or existing scripts somewhere?

My current not-really-a-script-yet test process so far is below:

```bash
# vsts-agent OSX install 'script' test
# Below represents a TEST scenario executed over SSH a line at a time with notes.

# install homebrew
/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"

# had to press enter to confirm - How to automate?
# had to enter password - probably fixed by running script as su?

brew cask install dotnet

# required one-time sudo password to create /usr/local/Caskroom
# maybe this can be done by script in advance?

# git --version
# already > 2.9.3
# May need to install git via brew on fresh machine?

brew install npm

mkdir /buildagent
cd /buildagent
# TODO: Check on ownership issue after context shifts to fresh machine script
sudo chown -R buildagent /buildagent

# -L = follow redirect
# TODO: How can we avoid hard coding this? or is it ok given automatic updates?
curl -L https://github.com/Microsoft/vsts-agent/releases/download/v2.109.1/vsts-agent-osx.10.11-x64-2.109.1.tar.gz > agent.tar.gz
tar zxvf agent.tar.gz
```

**Comment by TingluoHuang** [comment]
@ColinDabritzViewpoint 
Ideally, you should download the agent through the REST endpoint from your VisualStudio.com account. I don't think we have an endpoint to allow you download the latest version of a given agent package, I will add them later. for now, you can try https://github.com/Microsoft/vsts-agent/releases/latest, then grab the download url from the response.

**Comment by ColinDabritzViewpoint** [comment]
Ok, that's a couple approaches for now. What REST endpoint were you thinking of?

From these two URLS:
https://www.visualstudio.com/en-us/docs/integrate/api/overview
https://www.visualstudio.com/en-us/docs/integrate/api/contracts-page

I found endpoints for checking pool status and similar. Is there a 'Get latest agent' sort of endpoint I missed?

For now, I think getting it from the 'latest' page makes sense.

**Comment by TingluoHuang** [comment]
@ColinDabritzViewpoint I don't think we officially publish that endpoint. but it would be something like
`https://account.visualstudio.com/_apis/Distributedtask/packages/agent/latest?platform=win7-x64`

**Comment by ColinDabritzViewpoint** [comment]
Thanks! That seems to work, is there a platform list somewhere? I tried several osx variations with no luck..

**Comment by TingluoHuang** [comment]
@ColinDabritzViewpoint 
`https://account.visualstudio.com/_apis/Distributedtask/packages/agent?platform=osx.10.11-x64&$top=1`

**Comment by ColinDabritzViewpoint** [comment]
I went with the github URL as it is unauthenticated.

Thanks much for the pointers!

Here is the download and decompress Bash script I put together:

```bash
# Download Latest OSX VSTS Agent

LATEST_RELEASE_URL='https://github.com/Microsoft/vsts-agent/releases/latest'

# The above URL will redirect to the latest release page.
# We will parse the OSX download URL from that page, and retrieve the file.

# sed commands

# delete all lines that do NOT have 'href' followed by 'vsts-agent-osx'
# should match only the download anchor line
GET_OSX_URL='/href.*vsts-agent-osx/ !d'

# remove everything up to and including the first quote (by replacing with nothing)
DELETE_BEFORE_URL='s/[^"]*"//'

# removing everything after and including the closing quote (by replacing with nothing)
DELETE_AFTER_URL='s/".*//'

# add the base URL at the beginning of the line
PREPEND_BASE_URL='s,^,https://github.com,'

LATEST_OSX_AGENT_URL=`curl -Ls "$LATEST_RELEASE_URL" | \
sed -e "$GET_OSX_URL" -e "$DELETE_BEFORE_URL" -e "$DELETE_AFTER_URL" -e "$PREPEND_BASE_URL"`

echo Downloading latest agent from $LATEST_OSX_AGENT_URL

curl -L "$LATEST_OSX_AGENT_URL" -o vsts-agent-osx.10.11-x64.tar.gz

tar xvzf vsts-agent-osx.10.11-x64.tar.gz

```

**Comment by TingluoHuang** [comment]
Cool, glad you figure it out. i am going to close this issue.

**Comment by bryanmacfarlane** [comment]
Be really careful about "latest".  In GH, it's the latest one released time wise.  For example, on tuesday, I released the 2.108 agent.  On wednesday I released 2.105.3 from the 105 release branch.  105.3 is "latest".

Consider querying the VSTS accounts agent rest api which points to the latest url (what drives our download web UI).  That also considers our rollout process where we validate and update the urls scale unit by scale unit as we watch for issues.  That url represents "latest good"
